homeassistant/components/plum_lightpad/light.py

```diff
@@ -1,35 +1,65 @@
 """Support for Plum Lightpad lights."""
+import logging
+from typing import Callable, List
+
+from plumlightpad import Plum
+
 from homeassistant.components.light import (
     ATTR_BRIGHTNESS,
     ATTR_HS_COLOR,
     SUPPORT_BRIGHTNESS,
     SUPPORT_COLOR,
     LightEntity,
 )
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.entity import Entity
 import homeassistant.util.color as color_util
 
 from .const import DOMAIN
 
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: Callable[[List[Entity]], None],
+) -> None:
+    """Set up Plum Lightpad dimmer lights and glow rings."""
+
+    plum: Plum = hass.data[DOMAIN][entry.entry_id]
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
-    """Initialize the Plum Lightpad Light and GlowRing."""
-    if discovery_info is None:
-        return
+    def setup_entities(device) -> None:
+        entities = []
 
-    plum = hass.data[DOMAIN]
+        if "lpid" in device:
+            lightpad = plum.get_lightpad(device["lpid"])
+            entities.append(GlowRing(lightpad=lightpad))
 
-    entities = []
+        if "llid" in device:
+            logical_load = plum.get_load(device["llid"])
+            entities.append(PlumLight(load=logical_load))
 
-    if "lpid" in discovery_info:
-        lightpad = plum.get_lightpad(discovery_info["lpid"])
-        entities.append(GlowRing(lightpad=lightpad))
+        if entities:
+            async_add_entities(entities)
 
-    if "llid" in discovery_info:
-        logical_load = plum.get_load(discovery_info["llid"])
-        entities.append(PlumLight(load=logical_load))
+    async def new_load(device):
+        setup_entities(device)
 
-    if entities:
-        async_add_entities(entities)
+    async def new_lightpad(device):
+        setup_entities(device)
+
+    device_web_session = async_get_clientsession(hass, verify_ssl=False)
+    hass.loop.create_task(
+        plum.discover(
+            hass.loop,
+            loadListener=new_load,
+            lightpadListener=new_lightpad,
+            websession=device_web_session,
+        )
+    )
```

Comment:
I think the cleanup function from the config flow should be here. There is no need to cleanup until the actual devices are 'discovered' or hooked up.

Reply:
IMO, cleanup should be done by the code that 'owns' the instance, which is the code that creates it. The code in lighy.py does not own this instance, it simply gets pre-created instance from a shared structure.
Besides, it will only cleanup when HA goes down, not when cleanup is registered. Happy to chat about outside of PR if this is helpful. Thanks!